app/modules/net_info.c

```diff
@@ -0,0 +1,189 @@
+
+static ping_t nip_get( lua_State *L, int stack ) {
+  ping_t nip = (ping_t)luaL_checkudata(L, stack, "net_info.ping");
+  NODE_DBG("[nip_get] nip = %p\n", (void *) nip);
+  if (nip == NULL) {
+    return (ping_t)luaL_error(L, "ping object expected");
+  }
+  return nip;
+}
+  
+static void ping_received(void *arg, void *data) {
+    NODE_DBG("[ping_received] arg = %p, data = %p\n", arg, data);
+    ping_t nip  = (ping_t)arg; // override ping_option with net_info_ping_t to access the callback reference
+    struct ping_resp *pingresp = (struct ping_resp*)data; 
+
+    NODE_DBG("[ping_received] count = %d;arg: total_count=%d resp_time=%d seqno=%d timeout_count=%d bytes=%d total_bytes=%d total_time=%d ping_err=%d\n", nip->count, pingresp->total_count, pingresp->resp_time, pingresp->seqno, pingresp->timeout_count, pingresp->bytes, pingresp->total_bytes, pingresp->total_time, pingresp->ping_err);
+    
+    char ipaddrstr[16];
+    ip_addr_t source_ip;
+    source_ip.addr = nip->ping_opt.ip;
+    ipaddr_ntoa_r(&source_ip, ipaddrstr, sizeof(ipaddrstr));
+
+    lua_State *L = lua_getstate();
+    lua_rawgeti(L, LUA_REGISTRYINDEX, nip->ping_callback_ref);
+    lua_pushinteger(L, pingresp->bytes);
+    lua_pushstring(L, ipaddrstr);
+    lua_pushinteger(L, pingresp->seqno);
+    lua_pushinteger(L, pingresp->resp_time);
+    
+    (nip->count)--;
+    if (!nip->count) { // is this a last call?
+      if (nip->self_ref != LUA_REFNIL) {
+        NODE_DBG("[ping_received] luaL_unref nip->self_ref (%p, %p)\n", nip, nip->self_ref);
+        luaL_unref(L, LUA_REGISTRYINDEX, nip->self_ref);
+        nip->self_ref = LUA_NOREF;
+      }
+      if(nip->ping_callback_ref != LUA_NOREF) {
+        NODE_DBG("[ping_received] luaL_unref nip->ping_callback_ref (%p)\n", nip->ping_callback_ref);
+        luaL_unref(L, LUA_REGISTRYINDEX, nip->ping_callback_ref);
+        nip->ping_callback_ref = LUA_NOREF;
+      }
+    }
+    
+    lua_call(L, 4, 0);
+}
+
+static int net_info_ping(lua_State *L)
+{
+    NODE_DBG("[net_info_ping]\n");
+    ping_t nip = nip_get(L, 1);
+
+    if (nip->self_ref == LUA_NOREF) {
+        lua_pushvalue(L, 1);
+        nip->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+        NODE_DBG("[net_info_ping] nip = %p, nip->self_ref = %p\n", nip, nip->self_ref);
+    }
+    
+    const char *ping_target;
+    unsigned count = 4;
+     
+    // retrieve address arg (mandatory)
+    if (lua_isstring(L, 2)) {
+        ping_target = luaL_checkstring(L, 2);
+    } else {
+        return luaL_error(L, "no address specified");
+    }
+
+    // retrieve count arg (optional)
+    if (lua_isnumber(L, 3)) {
+        count = luaL_checkinteger(L, 3);
+    }
+   
+    // retrieve callback arg
+    NODE_DBG("[net_info_ping] nip->ping_callback_ref = %p\n", nip->ping_callback_ref);
+    if (nip->ping_callback_ref != LUA_NOREF) {
+      luaL_unref(L, LUA_REGISTRYINDEX, nip->ping_callback_ref);
+    }
+    nip->ping_callback_ref = LUA_NOREF;
```

Comment:
This could go inside the if brackets if I am not mistaken.

Reply:
done